### PR TITLE
Move ARM code to AArch32 sub-directory

### PR DESCRIPTION
**Pull request by carlocaione**
Preliminary refactoring work for #20263. Move the current ARM code to its own AArch32 sub-directory before introducing ARM64 code.

### CODE REVIEW

**Review by SebastianBoe — APPROVED** [review]

**Comment by ioannisg** [comment]
I think I've seen this failure before, during v2.1.0 stabilization, here: https://github.com/zephyrproject-rtos/zephyr/issues/20558

FYI
@galak @carlescufi @nashif,

**Comment by carlocaione** [comment]
> I think I've seen this failure before, during v2.1.0 stabilization, here: #20558
> 
> FYI
> @galak @carlescufi @nashif,

That is actually https://github.com/zephyrproject-rtos/zephyr/pull/21153#issuecomment-564941805

File: arch/arm/CMakeLists.txt [hunk 1]
```diff
@@ -1,26 +1,7 @@
+#
```
Comment (ioannisg) [inline arch/arm/CMakeLists.txt]: Not sure you need copyright header here, it's not added in arch cmakelists files. suggest to remove

File: CODEOWNERS [hunk 2]
```diff
@@ -16,9 +16,9 @@
 /.known-issues/                           @inakypg @nashif
 /arch/arc/                                @vonhust @ruuddw
 /arch/arm/                                @MaureenHelm @galak @ioannisg
-/arch/arm/core/cortex_m/cmse/             @ioannisg
-/arch/arm/include/cortex_m/cmse.h         @ioannisg
-/arch/arm/core/cortex_r/                  @MaureenHelm @galak @ioannisg @bbolen
+/arch/arm/core/aarch32/cortex_m/cmse/     @ioannisg
+/arch/arm/include/aarch32/cortex_m/cmse.h @ioannisg
+/arch/arm/core/aarch32/cortex_r/          @MaureenHelm @galak @ioannisg @bbolen
```
Comment (ioannisg) [inline CODEOWNERS]: suggest to add @stephanosio, too, for cortex-r

**Review by ioannisg — COMMENTED** [review]
Fine with me - some minor suggestions

**Comment by ioannisg** [comment]
So, AFAICS, you picked "aarch32" naming convention, and moved all Cortex-M, Cortex-R under this sub-directory; I suppose 32-bit Cortex-A will be there, too, if anyone wants to add it.

And "aarch64" will be the convention for 64-bit ARM (only Cortex-A, right?)

Funny, these names are describing execution states (in the ARM glossary) not architecture families.

But I haven't thought of better names, for these sub-directories...

**Comment by stephanosio** [comment]
> I suppose 32-bit Cortex-A will be there, too, if anyone wants to add it.

I am planning to add initial support for ARMv7-A after I finish stabilising the ARMv7-R port.

My current plan is to merge the (AArch32) Cortex-A and Cortex-R arch implementations into `arch/arm/aarch32/cortex_ar` directory as they are architecturally very similar; except MMU vs. MPU, of course.

> Funny, these names are describing execution states (in the ARM glossary) not architecture families.

IMO, this naming makes the most sense because an ARMv8-A processor, for example, can execute in both AArch32 and AArch64 states (i.e. if you have an ARMv8-A processor and want to execute in the AArch32 mode, you can still use the code under `arch/arm/aarch32`).

**Review by ioannisg — APPROVED** [review]

**Review by stephanosio — APPROVED** [review]

**Review by MaureenHelm — CHANGES_REQUESTED** [review]
I don't have an issue moving the current ARM code under AArch32, but I am concerned that changing the include paths will break out-of-tree boards and SoCs. I think we need to deprecate the old paths like we did in #16540.

**Comment by carlocaione** [comment]
> I don't have an issue moving the current ARM code under AArch32, but I am concerned that changing the include paths will break out-of-tree boards and SoCs. I think we need to deprecate the old paths like we did in #16540.

I personally do not like the idea that an open source project must take the burden to support out-of-tree code when the author of such code do not even bother to upstream the work and collaborate with the community. So I'm against this change.

That said, if that's really required to have this merged ASAP I can prepare a new respin.

**Comment by stephanosio** [comment]
@carlocaione On second thought, I also think public include path deprecation would be a good idea since the public arch headers under `include/arch/arm` are meant to be included by the applications.

Adding a new commit that adds the dummy headers under `include/arch/arm` that redirect to `include/arch/arm/aarch32` with a deprecation warning message would be nice.

> I personally do not like the idea that an open source project must take the burden to support out-of-tree code when the author of such code do not even bother to upstream the work and collaborate with the community

The "out-of-tree code" here can refer to anything from just the applications (both closed and open source, that do not make sense to be upstream-ed) to other projects that treat the upstream Zephyr repository as just a west module like any other (e.g. nRF Connect SDK, which is mostly open source).

**Comment by dbkinder** [comment]
The doc changes look OK, but I see lots of references to "arch/arm" in the board docs (as in boards/arm/nrf52840_pca10090/doc/index.rst).  Do these need updating too?

**Comment by carlocaione** [comment]
> The doc changes look OK, but I see lots of references to "arch/arm" in the board docs (as in boards/arm/nrf52840_pca10090/doc/index.rst). Do these need updating too?

The content of `arch/arm` was not changed by this PR. It still contains the same directories and files (+ `aarch32.cmake`) as before.

**Review by dbkinder — APPROVED** [review]
+1 for doc then, thanks for clarifying.

**Review by MaureenHelm — APPROVED** [review]

**Review by nashif — APPROVED** [review]

**Review by carlescufi — APPROVED** [review]

**Pull request merged**